// @triply/utils/lib/Constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.METADATA_IRI_FOR_CLASS_UNKNOWN = exports.RESOURCE_WIDGET_PATTERNS = exports.SUPPORTED_EXTENSIONS = exports.SPARQL_VARNAME_CHARS = exports.SUBSCRIPTIONS_PAGE = exports.NO_ACCOUNTS_LEFT_MESSAGE = exports.CANT_CREATE_ZERO_GRAPH_SERVICE_MESSAGE = exports.DUPLICATE_ASSETNAME_ERROR_MESSAGE = exports.GENERIC_FILE_UPLOAD_ERROR_MESSAGE = exports.AVATAR_SIZE_LIMIT = exports.SPARQL_RETRY_CONSTRUCT_SHRINK_FACTOR = exports.SPARQL_RESULT_DEFAULT_PAGE_SIZE = exports.SPARQL_RESULT_SET_MAX_ROWS = exports.FORWARD_PREDICATES_PAGE_SIZE = exports.BACKWARD_PREDICATES_PAGE_SIZE = exports.FORWARD_OBJECT_PAGE_SIZE = exports.BACKWARD_SUBJECTS_PAGE_SIZE = exports.MAX_GRAPHS_ON_LANDING_PAGE = exports.MAX_STORY_NAME_LENGTH = exports.MAX_QUERY_NAME_LENGTH = exports.MAX_SERVICE_NAME_LENGTH = exports.MAX_ACCOUNT_NAME_LENGTH = exports.MAX_DATASET_NAME_LENGTH = exports.MAX_EXAMPLE_RESOURCES = exports.QUERY_TYPES = exports.SERVICE_TYPES = exports.HDT_MAX_STATEMENTS = exports.HDT_MIN_STATEMENTS = void 0;
const path_1 = require("path");
const url = __importStar(require("url"));
exports.HDT_MIN_STATEMENTS = 1;
exports.HDT_MAX_STATEMENTS = 10000;
exports.SERVICE_TYPES = ["sparql", "sparql-jena", "elasticsearch"];
exports.QUERY_TYPES = ["sparql", "search"];
exports.MAX_EXAMPLE_RESOURCES = 30;
exports.MAX_DATASET_NAME_LENGTH = 40;
exports.MAX_ACCOUNT_NAME_LENGTH = 40;
exports.MAX_SERVICE_NAME_LENGTH = 40;
exports.MAX_QUERY_NAME_LENGTH = 40;
exports.MAX_STORY_NAME_LENGTH = 40;
exports.MAX_GRAPHS_ON_LANDING_PAGE = 10;
exports.BACKWARD_SUBJECTS_PAGE_SIZE = 40;
exports.FORWARD_OBJECT_PAGE_SIZE = 10;
exports.BACKWARD_PREDICATES_PAGE_SIZE = 10;
exports.FORWARD_PREDICATES_PAGE_SIZE = 50;
exports.SPARQL_RESULT_SET_MAX_ROWS = 10000;
exports.SPARQL_RESULT_DEFAULT_PAGE_SIZE = 100;
exports.SPARQL_RETRY_CONSTRUCT_SHRINK_FACTOR = 0.8;
exports.AVATAR_SIZE_LIMIT = "5mb";
exports.GENERIC_FILE_UPLOAD_ERROR_MESSAGE = "An error occurred while uploading the file.";
exports.DUPLICATE_ASSETNAME_ERROR_MESSAGE = "An asset with this name has already been uploaded.";
exports.CANT_CREATE_ZERO_GRAPH_SERVICE_MESSAGE = "Cannot create a service for a dataset without graphs.";
exports.NO_ACCOUNTS_LEFT_MESSAGE = "There are currently no more available accounts.";
exports.SUBSCRIPTIONS_PAGE = "https://triply.cc/subscriptions";
exports.SPARQL_VARNAME_CHARS = "[0-9a-zA-Z_\u00C0-\u00D6\u00D8-\u00F6\u0370-\u037d\u037F-\u1FFF\u200C-\u200D\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD]";
exports.SUPPORTED_EXTENSIONS = [
    "gz", "bz2", "xz", "tar", "zip", "tgz",
    "rdf",
    "rdfs",
    'owx',
    'owl',
    "nt",
    "ttl",
    "nq",
    "trig",
    "n3"
];
const imageExtensions = [".jpg", ".jpeg", ".png", ".gif", ".svg", ".bmp", ".tiff"];
exports.RESOURCE_WIDGET_PATTERNS = {
    IMAGE_PATTERNS: [
        ["http://xmlns.com/foaf/0.1/depiction", null, "http://xmlns.com/foaf/0.1/thumbnail", null],
        ["https://schema.org/image", null, "https://schema.org/contentUrl"],
        ["https://schema.org/thumbnail", null, "https://schema.org/contentUrl"],
        ["http://www.w3.org/1999/02/22-rdf-syntax-ns#type", { value: "https://schema.org/ImageObject" }],
        ["http://xmlns.com/foaf/0.1/depiction", null],
        ["http://xmlns.com/foaf/0.1/thumbnail", null],
        ["https://schema.org/image"],
        [null, { datatype: "https://triply.cc/triply/def/imageURI" }],
        [
            null,
            {
                datatype: "http://www.w3.org/2001/XMLSchema#anyURI",
                validationFunction: (term) => imageExtensions.indexOf(path_1.extname(url.parse(term.value).pathname || "").toLowerCase()) >= 0 ||
                    imageExtensions.indexOf(path_1.extname(url.parse(term.value).search || "").toLowerCase()) >= 0,
            },
        ],
    ],
    LABEL_PATTERNS: [
        ["http://www.w3.org/2000/01/rdf-schema#label", { language: "en" }],
        ["http://www.w3.org/2000/01/rdf-schema#label", undefined],
        ["http://xmlns.com/foaf/0.1/name", { language: "en" }],
        ["http://xmlns.com/foaf/0.1/name", undefined],
        ["https://schema.org/name", { language: "en" }],
        ["https://schema.org/name", undefined],
        ["http://www.w3.org/2004/02/skos/core#prefLabel", { language: "en" }],
        ["http://www.w3.org/2004/02/skos/core#prefLabel", undefined],
        ["http://purl.org/dc/terms/title", { language: "en" }],
        ["http://purl.org/dc/terms/title", undefined],
        ["http://purl.org/dc/elements/1.1/title", { language: "en" }],
        ["http://purl.org/dc/elements/1.1/title", undefined],
    ],
    GEO_PATTERNS: [
        ["http://www.opengis.net/ont/geosparql#asWKT", undefined],
        ["http://www.opengis.net/ont/geosparql#hasGeometry", undefined, "http://www.opengis.net/ont/geosparql#asWKT"],
    ],
    LAT_PATTERNS: [["http://www.w3.org/2003/01/geo/wgs84_pos#lat", undefined]],
    LONG_PATTERNS: [["http://www.w3.org/2003/01/geo/wgs84_pos#long", undefined]],
    AUDIO_PATTERNS: [
        ["https://schema.org/audio"],
        ["http://www.w3.org/1999/02/22-rdf-syntax-ns#type", { value: "https://schema.org/AudioObject" }],
    ],
    VIDEO_PATTERNS: [
        ["https://schema.org/video"],
        ["http://www.w3.org/1999/02/22-rdf-syntax-ns#type", { value: "https://schema.org/VideoObject" }],
    ],
    MEDIA_PATTERNS: [["https://schema.org/contentUrl"]],
    MEDIA_ENCODING_PATTERNS: [["https://schema.org/encodingFormat"]],
};
exports.METADATA_IRI_FOR_CLASS_UNKNOWN = "UNKNOWN:unknown";
