// @triply/utils/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Constants = exports.Routes = exports.Models = exports.nTriply = exports.prefixUtils = exports.LiteralUtils = exports.geo = exports.getEnvironmentVarsAsObject = exports.ConfigTemplate = void 0;
var ConfigTemplate_1 = require("./ConfigTemplate");
Object.defineProperty(exports, "ConfigTemplate", { enumerable: true, get: function () { return __importDefault(ConfigTemplate_1).default; } });
Object.defineProperty(exports, "getEnvironmentVarsAsObject", { enumerable: true, get: function () { return ConfigTemplate_1.getEnvironmentVarsAsObject; } });
const _prefixUtils = __importStar(require("./prefixUtils"));
const _nTriply = __importStar(require("./nTriply"));
const _literalUtils = __importStar(require("./literalUtils"));
const _geo = __importStar(require("./geo"));
exports.geo = _geo;
exports.LiteralUtils = _literalUtils;
exports.prefixUtils = _prefixUtils;
exports.nTriply = _nTriply;
const Models = __importStar(require("./Models"));
exports.Models = Models;
const Routes = __importStar(require("./Routes"));
exports.Routes = Routes;
const Constants = __importStar(require("./Constants"));
exports.Constants = Constants;
